text/0000-unicode-string.md

```diff
@@ -0,0 +1,97 @@
+
+Change Writer in buffered/writer.pony to accept Encoder parameters in the write() and writev() functions. 
+
+Add a FileCharacters class in buffered/file_characters.pony that provides an iterator of characters in a file. The implementation will be similar to the FileLines class.
+
+Change character literals so that a character literal can only represent a single unicode codepoint. The following would be valid character literals:
+```
+let l1 = 'A'
+let l2 = '🐎'
+let l3 = '\x61' \\ only hex values up to 7F will be accepted.
+let l4 = '\u20AC' \\ Unicode codepoint '€'
+let l5 = '\U01F3A0 \\ Unicode codepoint '🎠'
+```
+
+Change the Pony tutorial to reflect the changes to the String class and character literals. Also state clearly that UTF-8 is the only valid encoding for Pony source code.
+
+Not Supported:
+1. lower() and upper() for unicode characters. Should remove lower_in_place() and upper_in_place() because these conversion are not safe.
+1. The StdStream's (out, err, in) do not support specifying a character encoding. Ideally, the encoding for these streams would be set by the system at run-time based on the default encoding of the underlying system. For now, they will use utf-8 only.
+
+# How We Teach This
+
+This can be presented as a continuation of existing Pony patterns.
+
+The Pony tutorial will need to be updated to reflect these changes.
+
+# How We Test This
+
+Extend CI coverage to cover these changes and new features.
+
+# Drawbacks
+
+This is a change to the String API, and as such will break existing programs. String functions returning or taking as parameters U8 values now returning or taking U32 values will probably be the most common source of program breakage. Also, programs that used the existing unicode friendly functions in String will need to change, but they should be greatly simplified.
+
+# Alternatives
+
+1. Leave the String class as it is. This is likely to result in hidden errors in many programs that work with Strings as they will not work correctly with unicode data if they encounter it. It will also make it difficult to use Pony in settings where ASCII is not sufficient for local language (most non English speaking countries).
+1. A more complicated implementation with multiple String types capable of storing text data internally using different byte encodings. This approach would improve performance in situations where strings needed to be created from bytes of various encodings. The String type could be matched to the native byte encoding to eliminate any need for conversion. Such an implementation would add complexity to the String API as it would require use of a String trait with multiple implementations. It would also add considerable complexity to the String class implementation.
```

Comment:
[As I mentioned in our discussion in Zulip](https://ponylang.zulipchat.com/#narrow/stream/192795-contribute-to.20Pony/topic/Changes.20to.20String.20class/near/202574862), I would prefer to see a version of this RFC that takes the encoding as a type parameter to all encoding-aware operations within the `String` class, using UTF-8 as the default (so that everything in this RFC works for the user as you describe by default), but so that non-UTF-8 use cases can also be supported.

This does not necessarily mean you need multiple String classes or a String trait - just multiple encoding classes and an encoding trait (I believe you already have an encoder/decoder traits as part of this RFC, so that could potentially be used).